semgrep-core/toy-matcher/Matcher.ml

```diff
@@ -0,0 +1,562 @@
+(*
+   A matcher operating on sequences of symbols,
+   using the same basic matching algorithm as semgrep-core.
+
+   The goals are:
+
+   - Clarify semgrep-core's matching algorithm by not having to deal with any
+     of the real-world complexity.
+   - Keep the semgrep-core implementation separate and easy to follow,
+     without functors.
+*)
+
+open Printf
+
+type pattern_atom =
+  | Any_symbol         (* match any single symbol *)
+  | Symbol of char     (* match a specific symbol *)
+  | Ellipsis           (* match any sequence of symbols *)
+  | Backref of string  (* match the same thing that was previous matched
+                          by the pattern atom of that name. *)
+
+(*
+   A pattern is a list of pattern atoms with an optional name.
+   The presence of a name indicates that the symbol or sequence it matches
+   must be captured, i.e. stored in the environment.
+*)
+type pattern = (pattern_atom * string option) list
+
+type input = char list
+
+(* Captured subsequences. *)
+module Env = Map.Make (String)
+type env = char list Env.t
+
+(*
+   Accumulator use to store the symbols matched by the current ellipsis.
+
+     None = we're not in an ellipsis
+     Some None = we're in an ellipsis but it has no name and it's not capturing
+
+     Some (Some ("foo", acc)) = acc is the stack of symbols, i.e. the list
+     of symbols captured so far by the current ellipsis in reverse order
+*)
+type ellipsis = (string * char list) option option
+
+type stat = {
+  mutable match_calls: int;
+}
+
+let sample_pattern : pattern = [
+  Symbol 'A', None;
+  Any_symbol, Some "thing";
+  Ellipsis, None;
+  Backref "thing", None;
+]
+
+(*
+   Unlike in semgrep syntax, we distinguish metavariable assignments
+   from metavariable dereferencing.
+
+   This function checks that variable assignments are unique and that
+   backreferences refer to a valid variable name.
+
+   This is a sanity check, not part of the matching algorithm.
+*)
+let rec check_pattern_ env (pat : pattern) =
+  match pat with
+  | [] -> ()
+  | (atom, opt_name) :: pat ->
+      let new_env =
+        match opt_name with
+        | None -> env
+        | Some name ->
+            if List.mem name env then
+              failwith ("multiple atoms have the same name: " ^ name)
+            else
+              name :: env
+      in
+      (match atom with
+       | Any_symbol -> ()
+       | Symbol _ -> ()
+       | Ellipsis -> ()
+       | Backref name ->
+           if not (List.mem name env) then
+             failwith ("backreference to invalid name: " ^ name)
+      );
+      check_pattern_ new_env pat
+
+let check_pattern pat =
+  check_pattern_ [] pat
+
+(*
+   Initialize the capture of symbols by an ellipsis.
+   'init_acc' is a stack, holding symbols in reverse order.
+*)
+let init_ellipsis opt_name init_acc =
```

Comment:
this would be clearer if we had 3 constructors instead of Some Some.

Reply:
I did that.